yarpc-core/src/main/java/io/yarpc/transport/tchannel/TChannelServerInboundHandler.java

```
@@ -0,0 +1,140 @@
+import com.uber.tchannel.messages.Request;
+import com.uber.tchannel.messages.Response;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.util.AttributeKey;
+import io.yarpc.HandlerRegistry;
+import io.yarpc.encoding.Encoding;
+import io.yarpc.exception.ExceptionMessages;
+import io.yarpc.exception.InvalidInboundRequestException;
+import io.yarpc.exception.InvalidOutboundResponseException;
+import io.yarpc.exception.ServerException;
+import io.yarpc.handler.InboundHandler;
+import io.yarpc.util.HeaderUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+
+@ChannelHandler.Sharable
+public class TChannelServerInboundHandler extends SimpleChannelInboundHandler<Request> {
+
+    private final static Logger logger = LoggerFactory.getLogger(TChannelServerInboundHandler.class);
+    private final static AttributeKey<Long> REQUEST_ID = AttributeKey.valueOf("request_id");
+    private final HandlerRegistry handlerRegistry;
+
+    public TChannelServerInboundHandler(HandlerRegistry handlerRegistry) {
+        this.handlerRegistry = handlerRegistry;
+    }
+
+    public static ByteBuf encodeHeaderForTChannel(io.yarpc.Request<ByteBuffer> transportRequest, io.yarpc
+            .Response<ByteBuffer> transportResponse) throws InvalidOutboundResponseException {
+        ByteBuf headers = null;
+        try {
+            if (transportRequest.getEncoding().equals(Encoding.JSON)) {
+                headers = HeaderUtil.encodeHeadersJsonStyle(transportResponse.getHeaders());
+            } else {
+                headers = HeaderUtil.encodeHeaders(transportResponse.getHeaders());
+            }
+            return headers;
+        } catch (Exception e) {
+            throw new InvalidOutboundResponseException(
```

What's with these catch-all exceptions everywhere? `encodeHeaders` raises nothing and `encodeJsonStyle` raises only `JsonProcessingException`. Be specific about the exceptions you're catching where possible. Encoding and decoding should have a very narrow set of possible exceptions.